// @dynamic-labs/stellar/src/connectors/LobstrWalletConnector/LobstrWalletConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.cjs');
var signerExtensionApi = require('@lobstrco/signer-extension-api');
var StellarWalletConnector = require('../StellarWalletConnector/StellarWalletConnector.cjs');
var StellarWallet = require('../../wallet/StellarWallet.cjs');
require('@stellar/stellar-sdk');
var getNetworkFromAddress = require('../../utils/getNetworkFromAddress.cjs');
require('@dynamic-labs/utils');

/**
 * Lobstr Wallet provider adapter that implements IStellarProvider
 * using the Lobstr signer extension API.
 */
const createLobstrProvider = () => ({
    // Lobstr SDK doesn't have a close method
    close: () => undefined,
    connect: () => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const publicKey = yield signerExtensionApi.getPublicKey();
        return publicKey;
    }),
    // Lobstr SDK doesn't have a disconnect method
    disconnect: () => _tslib.__awaiter(void 0, void 0, void 0, function* () { return undefined; }),
    // Lobstr uses mainnet by default
    getNetwork: () => _tslib.__awaiter(void 0, void 0, void 0, function* () { return 'PUBLIC'; }),
    isUnlocked: () => _tslib.__awaiter(void 0, void 0, void 0, function* () { return signerExtensionApi.isConnected(); }),
    // Lobstr SDK doesn't support event subscriptions
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    on: (event, callback) => undefined,
    sign: (transactionXdr) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const signedXdr = yield signerExtensionApi.signTransaction(transactionXdr);
        return signedXdr;
    }),
    signMessage: (message) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const response = yield signerExtensionApi.signMessage(message);
        return (_a = response === null || response === void 0 ? void 0 : response.signedMessage) !== null && _a !== void 0 ? _a : '';
    }),
});
class LobstrWalletConnector extends StellarWalletConnector.StellarWalletConnector {
    constructor(opts) {
        super('Lobstr', opts);
        this.name = 'Lobstr';
        this.ChainWallet = StellarWallet.StellarWallet;
        this.connectedChain = 'STELLAR';
        this.supportedChains = ['STELLAR'];
        this.provider = createLobstrProvider();
    }
    static isConnected() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const connected = yield signerExtensionApi.isConnected();
            LobstrWalletConnector.isInstalled = connected;
            return connected;
        });
    }
    getProvider() {
        if (!this.isInstalledOnBrowser()) {
            return undefined;
        }
        if (!this.provider) {
            this.provider = createLobstrProvider();
        }
        return this.provider;
    }
    getNetwork() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const address = yield this.getAddress();
            if (!address) {
                return undefined;
            }
            if (!this.stellarNetworks.length) {
                return undefined;
            }
            const network = yield getNetworkFromAddress.getNetworkFromAddress(address, this.stellarNetworks[0]);
            return network;
        });
    }
    isInstalledOnBrowser() {
        return LobstrWalletConnector.isInstalled;
    }
    /**
     * Lobstr uses its own SDK method to fetch the public key
     * instead of the provider's connect method.
     */
    fetchPublicAddress() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return signerExtensionApi.getPublicKey();
        });
    }
}
LobstrWalletConnector.isInstalled = false;

exports.LobstrWalletConnector = LobstrWalletConnector;
