// @dynamic-labs/stellar/src/connectors/LobstrWalletConnector/LobstrWalletConnector.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { getPublicKey, isConnected, signTransaction, signMessage } from '@lobstrco/signer-extension-api';
import { StellarWalletConnector } from '../StellarWalletConnector/StellarWalletConnector.js';
import { StellarWallet } from '../../wallet/StellarWallet.js';
import '@stellar/stellar-sdk';
import { getNetworkFromAddress } from '../../utils/getNetworkFromAddress.js';
import '@dynamic-labs/utils';

/**
 * Lobstr Wallet provider adapter that implements IStellarProvider
 * using the Lobstr signer extension API.
 */
const createLobstrProvider = () => ({
    // Lobstr SDK doesn't have a close method
    close: () => undefined,
    connect: () => __awaiter(void 0, void 0, void 0, function* () {
        const publicKey = yield getPublicKey();
        return publicKey;
    }),
    // Lobstr SDK doesn't have a disconnect method
    disconnect: () => __awaiter(void 0, void 0, void 0, function* () { return undefined; }),
    // Lobstr uses mainnet by default
    getNetwork: () => __awaiter(void 0, void 0, void 0, function* () { return 'PUBLIC'; }),
    isUnlocked: () => __awaiter(void 0, void 0, void 0, function* () { return isConnected(); }),
    // Lobstr SDK doesn't support event subscriptions
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    on: (event, callback) => undefined,
    sign: (transactionXdr) => __awaiter(void 0, void 0, void 0, function* () {
        const signedXdr = yield signTransaction(transactionXdr);
        return signedXdr;
    }),
    signMessage: (message) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const response = yield signMessage(message);
        return (_a = response === null || response === void 0 ? void 0 : response.signedMessage) !== null && _a !== void 0 ? _a : '';
    }),
});
class LobstrWalletConnector extends StellarWalletConnector {
    constructor(opts) {
        super('Lobstr', opts);
        this.name = 'Lobstr';
        this.ChainWallet = StellarWallet;
        this.connectedChain = 'STELLAR';
        this.supportedChains = ['STELLAR'];
        this.provider = createLobstrProvider();
    }
    static isConnected() {
        return __awaiter(this, void 0, void 0, function* () {
            const connected = yield isConnected();
            LobstrWalletConnector.isInstalled = connected;
            return connected;
        });
    }
    getProvider() {
        if (!this.isInstalledOnBrowser()) {
            return undefined;
        }
        if (!this.provider) {
            this.provider = createLobstrProvider();
        }
        return this.provider;
    }
    getNetwork() {
        return __awaiter(this, void 0, void 0, function* () {
            const address = yield this.getAddress();
            if (!address) {
                return undefined;
            }
            if (!this.stellarNetworks.length) {
                return undefined;
            }
            const network = yield getNetworkFromAddress(address, this.stellarNetworks[0]);
            return network;
        });
    }
    isInstalledOnBrowser() {
        return LobstrWalletConnector.isInstalled;
    }
    /**
     * Lobstr uses its own SDK method to fetch the public key
     * instead of the provider's connect method.
     */
    fetchPublicAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            return getPublicKey();
        });
    }
}
LobstrWalletConnector.isInstalled = false;

export { LobstrWalletConnector };
