// @dynamic-labs/stellar/src/connectors/OneKeyWalletConnector/OneKeyWalletConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var StellarWalletConnector = require('../StellarWalletConnector/StellarWalletConnector.cjs');
var OneKeyProvider = require('./OneKeyProvider.cjs');

/**
 * OneKey Wallet connector for Stellar blockchain.
 * Implements Stellar wallet functionality using the OneKey browser extension.
 */
class OneKeyWalletConnector extends StellarWalletConnector.StellarWalletConnector {
    constructor(opts) {
        super('OneKey', opts);
        this.name = 'OneKey';
        this.overrideKey = 'onekeystellar';
    }
    getProvider() {
        if (!this.isInstalledOnBrowser()) {
            return undefined;
        }
        if (!this.provider) {
            try {
                this.provider = OneKeyProvider.createOneKeyProvider();
            }
            catch (error) {
                // Provider not available
                return undefined;
            }
        }
        return this.provider;
    }
    isInstalledOnBrowser() {
        var _a;
        if (typeof window === 'undefined') {
            return false;
        }
        return ((_a = window.$onekey) === null || _a === void 0 ? void 0 : _a.stellar) !== undefined;
    }
}

exports.OneKeyWalletConnector = OneKeyWalletConnector;
