// @dynamic-labs/stellar/src/wallet/StellarWallet.js
'use client'
import { __awaiter } from '../../_virtual/_tslib.js';
import { TransactionBuilder } from '@stellar/stellar-sdk';
import { Wallet } from '@dynamic-labs/wallet-connector-core';
import { buildPaymentTransaction } from '../utils/buildPaymentTransaction.js';
import { checkTrustline } from '../utils/checkTrustline.js';
import { createPaymentAsset } from '../utils/createPaymentAsset.js';
import { getAccountBalance } from '../utils/getAccountBalance.js';
import '@dynamic-labs/utils';

/**
 * Stellar wallet implementation that provides chain-specific functionality
 * for interacting with the Stellar blockchain.
 *
 * This class extends the base Wallet class and provides Stellar-specific
 * methods for transactions, signing, and network operations.
 */
class StellarWallet extends Wallet {
    /**
     * Get balance of the wallet.
     * @param assetCode - Optional asset code. If not provided, returns native XLM balance
     * @param assetIssuer - Optional asset issuer. Required if assetCode is provided
     * @returns Balance of the wallet
     */
    getBalance(assetCode, assetIssuer) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._connector.connect();
            const horizonServer = yield this.getHorizonServer();
            try {
                const account = yield horizonServer.loadAccount(this.address);
                return getAccountBalance(account, assetCode, assetIssuer);
            }
            catch (error) {
                return '0';
            }
        });
    }
    /**
     * Sends balance to another address.
     * Supports both native XLM and custom Stellar assets.
     *
     * @param params - Send balance parameters
     * @param params.amount - Amount to send
     * @param params.toAddress - Recipient address (Stellar public key)
     * @param params.token - Optional token information for non-XLM transfers
     * @param params.token.address - Token address in format "CODE:ISSUER"
     * @param params.token.decimals - Token decimals (optional, Stellar uses 7 decimals by default)
     * @returns Transaction hash
     */
    sendBalance(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { amount, toAddress, token } = params;
            yield this._connector.connect();
            const horizonServer = yield this.getHorizonServer();
            const sourceAccount = yield horizonServer.loadAccount(this.address);
            const networkPassphrase = yield this._connector.getNetworkPassphrase();
            let asset;
            if (token === null || token === void 0 ? void 0 : token.address) {
                const [code, issuer] = token.address.split(/[:-]/);
                if (!code || !issuer) {
                    throw new Error('Invalid token address format. Expected "CODE:ISSUER" or "CODE-ISSUER" (e.g., "USDC:GXXXXXXX..." or "USDC-GXXXXXXX...")');
                }
                asset = { code, issuer };
            }
            const paymentAsset = createPaymentAsset(asset);
            let destinationAccount;
            try {
                destinationAccount = yield horizonServer.loadAccount(toAddress);
            }
            catch (error) {
                throw new Error(`Failed to load destination account ${toAddress}. The account may not exist or be funded.`);
            }
            // Check if destination account has trustline for non-native assets
            if (!paymentAsset.isNative()) {
                const hasTrustline = checkTrustline(destinationAccount, paymentAsset);
                if (!hasTrustline) {
                    const assetCode = paymentAsset.getCode();
                    const assetIssuer = paymentAsset.getIssuer();
                    throw new Error(`Destination account ${toAddress} does not have a trustline for asset ${assetCode}:${assetIssuer}. ` +
                        'The recipient must establish a trustline before receiving this asset.');
                }
            }
            const transaction = buildPaymentTransaction({
                amount,
                asset: paymentAsset,
                memo: undefined,
                networkPassphrase,
                sourceAccount,
                toAddress,
            });
            const transactionXdr = transaction.toXDR();
            const signedXdr = yield this.signTransaction(transactionXdr);
            const transactionToSubmit = TransactionBuilder.fromXDR(signedXdr, networkPassphrase);
            const response = yield horizonServer.submitTransaction(transactionToSubmit);
            return response.hash;
        });
    }
    /**
     * Signs a Stellar transaction XDR.
     *
     * @param transactionXdr - The transaction XDR to sign
     * @returns The signed transaction XDR
     */
    signTransaction(transactionXdr) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._connector.connect();
            return this._connector.signTransaction(transactionXdr);
        });
    }
    /**
     * Signs a message.
     *
     * @param message - The message to sign
     * @returns The signature
     */
    signMessage(message) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._connector.connect();
            const result = yield this._connector.signMessage(message);
            if (!result) {
                throw new Error('Failed to sign message');
            }
            return result;
        });
    }
    /**
     * Returns the wallet's current network.
     *
     * @returns The current network or undefined if not available
     */
    getNetworkInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getNetwork();
        });
    }
    /**
     * Retrieves the Horizon Server instance for interacting with the Stellar network.
     *
     * @returns A promise that resolves to the Horizon.Server instance
     */
    getHorizonServer() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getHorizonServer();
        });
    }
}

export { StellarWallet };
