// @dynamic-labs/sdk-api-core/src/models/TokenScope.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Valid scopes for an elevated access token
 * @export
 * @enum {string}
 */
exports.TokenScope = void 0;
(function (TokenScope) {
    TokenScope["Walletexport"] = "wallet:export";
})(exports.TokenScope || (exports.TokenScope = {}));
function TokenScopeFromJSON(json) {
    return TokenScopeFromJSONTyped(json);
}
function TokenScopeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function TokenScopeToJSON(value) {
    return value;
}

exports.TokenScopeFromJSON = TokenScopeFromJSON;
exports.TokenScopeFromJSONTyped = TokenScopeFromJSONTyped;
exports.TokenScopeToJSON = TokenScopeToJSON;
