// @dynamic-labs/stellar/src/index.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assertPackageVersion = require('@dynamic-labs/assert-package-version');
var _package = require('../package.cjs');
var fetchInjectedWalletConnectors = require('./injected/fetchInjectedWalletConnectors.cjs');
var StellarWallet = require('./wallet/StellarWallet.cjs');
var isStellarWallet = require('./wallet/isStellarWallet/isStellarWallet.cjs');
var StellarWalletConnector = require('./connectors/StellarWalletConnector/StellarWalletConnector.cjs');
var StellarLocalStorageCache = require('./StellarLocalStorageCache.cjs');
require('@stellar/stellar-sdk');
var getNetworkFromAddress = require('./utils/getNetworkFromAddress.cjs');
require('../_virtual/_tslib.cjs');
require('@dynamic-labs/utils');

assertPackageVersion.assertPackageVersion('@dynamic-labs/stellar', _package.version);
// Wallet connector factory (to be implemented by specific wallet connectors)
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const StellarWalletConnectors = (_props) => [
    ...fetchInjectedWalletConnectors.fetchInjectedWalletConnectors(),
    // Specific wallet connectors will be added here by implementers
];

exports.StellarWallet = StellarWallet.StellarWallet;
exports.isStellarWallet = isStellarWallet.isStellarWallet;
exports.StellarWalletConnector = StellarWalletConnector.StellarWalletConnector;
exports.StellarLocalStorageCache = StellarLocalStorageCache.StellarLocalStorageCache;
exports.getNetworkFromAddress = getNetworkFromAddress.getNetworkFromAddress;
exports.StellarWalletConnectors = StellarWalletConnectors;
