// @dynamic-labs/stellar/src/index.js
'use client'
import { assertPackageVersion } from '@dynamic-labs/assert-package-version';
import { version } from '../package.js';
import { fetchInjectedWalletConnectors } from './injected/fetchInjectedWalletConnectors.js';
export { StellarWallet } from './wallet/StellarWallet.js';
export { isStellarWallet } from './wallet/isStellarWallet/isStellarWallet.js';
export { StellarWalletConnector } from './connectors/StellarWalletConnector/StellarWalletConnector.js';
export { StellarLocalStorageCache } from './StellarLocalStorageCache.js';
import '@stellar/stellar-sdk';
export { getNetworkFromAddress } from './utils/getNetworkFromAddress.js';
import '../_virtual/_tslib.js';
import '@dynamic-labs/utils';

assertPackageVersion('@dynamic-labs/stellar', version);
// Wallet connector factory (to be implemented by specific wallet connectors)
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const StellarWalletConnectors = (_props) => [
    ...fetchInjectedWalletConnectors(),
    // Specific wallet connectors will be added here by implementers
];

export { StellarWalletConnectors };
