// @dynamic-labs/stellar/src/StellarLocalStorageCache.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../_virtual/_tslib.cjs');
var utils = require('@dynamic-labs/utils');

class StellarLocalStorageCache {
    constructor(key) {
        this.CONNECTED_ACCOUNTS_KEY = `stellar_${key}_connectedAccounts`;
    }
    getActiveAccount() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield utils.getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            if (!connectedAccounts) {
                return;
            }
            const activeAddress = Object.keys(connectedAccounts).find((key) => connectedAccounts[key].active);
            if (!activeAddress) {
                return;
            }
            return {
                additionalAddresses: connectedAccounts[activeAddress].additionalAddresses,
                address: activeAddress,
            };
        });
    }
    getConnectedAccounts() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield utils.getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            return connectedAccounts;
        });
    }
    getConnectedAccount(address) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield utils.getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            return connectedAccounts === null || connectedAccounts === void 0 ? void 0 : connectedAccounts[address];
        });
    }
    setConnectedAccount(address, account) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            let connectedAccounts = yield utils.getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            if (!connectedAccounts) {
                connectedAccounts = {};
            }
            if (account.active) {
                Object.keys(connectedAccounts).forEach((key) => {
                    if (connectedAccounts) {
                        connectedAccounts[key].active = false;
                    }
                });
            }
            connectedAccounts[address] = account;
            return utils.setItemAsync(this.CONNECTED_ACCOUNTS_KEY, connectedAccounts);
        });
    }
    clearConnectedAcccounts() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return utils.removeItemAsync(this.CONNECTED_ACCOUNTS_KEY);
        });
    }
}

exports.StellarLocalStorageCache = StellarLocalStorageCache;
