// @dynamic-labs/stellar/src/StellarLocalStorageCache.js
'use client'
import { __awaiter } from '../_virtual/_tslib.js';
import { getItemAsync, setItemAsync, removeItemAsync } from '@dynamic-labs/utils';

class StellarLocalStorageCache {
    constructor(key) {
        this.CONNECTED_ACCOUNTS_KEY = `stellar_${key}_connectedAccounts`;
    }
    getActiveAccount() {
        return __awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            if (!connectedAccounts) {
                return;
            }
            const activeAddress = Object.keys(connectedAccounts).find((key) => connectedAccounts[key].active);
            if (!activeAddress) {
                return;
            }
            return {
                additionalAddresses: connectedAccounts[activeAddress].additionalAddresses,
                address: activeAddress,
            };
        });
    }
    getConnectedAccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            return connectedAccounts;
        });
    }
    getConnectedAccount(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const connectedAccounts = yield getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            return connectedAccounts === null || connectedAccounts === void 0 ? void 0 : connectedAccounts[address];
        });
    }
    setConnectedAccount(address, account) {
        return __awaiter(this, void 0, void 0, function* () {
            let connectedAccounts = yield getItemAsync(this.CONNECTED_ACCOUNTS_KEY);
            if (!connectedAccounts) {
                connectedAccounts = {};
            }
            if (account.active) {
                Object.keys(connectedAccounts).forEach((key) => {
                    if (connectedAccounts) {
                        connectedAccounts[key].active = false;
                    }
                });
            }
            connectedAccounts[address] = account;
            return setItemAsync(this.CONNECTED_ACCOUNTS_KEY, connectedAccounts);
        });
    }
    clearConnectedAcccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            return removeItemAsync(this.CONNECTED_ACCOUNTS_KEY);
        });
    }
}

export { StellarLocalStorageCache };
