// @dynamic-labs/stellar/src/connectors/FreighterWalletConnector/FreighterWalletConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.cjs');
var freighterApi = require('@stellar/freighter-api');
var StellarWalletConnector = require('../StellarWalletConnector/StellarWalletConnector.cjs');
var FreighterProvider = require('./FreighterProvider.cjs');

/**
 * Freighter Wallet connector for Stellar blockchain.
 * Implements Stellar wallet functionality using the Freighter browser extension.
 */
class FreighterWalletConnector extends StellarWalletConnector.StellarWalletConnector {
    constructor(opts) {
        super('Freighter', opts);
        this.name = 'Freighter';
        this.overrideKey = 'freighter';
    }
    getProvider() {
        if (!this.isInstalledOnBrowser()) {
            return undefined;
        }
        if (!this.provider) {
            this.provider = FreighterProvider.createFreighterProvider();
        }
        return this.provider;
    }
    isInstalledOnBrowser() {
        return FreighterWalletConnector.isInstalled;
    }
    static isConnected() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const response = yield freighterApi.isConnected();
            const connected = (_a = response === null || response === void 0 ? void 0 : response.isConnected) !== null && _a !== void 0 ? _a : false;
            FreighterWalletConnector.isInstalled = connected;
            return connected;
        });
    }
}
FreighterWalletConnector.isInstalled = false;

exports.FreighterWalletConnector = FreighterWalletConnector;
