// @dynamic-labs/stellar/src/connectors/FreighterWalletConnector/FreighterWalletConnector.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { isConnected } from '@stellar/freighter-api';
import { StellarWalletConnector } from '../StellarWalletConnector/StellarWalletConnector.js';
import { createFreighterProvider } from './FreighterProvider.js';

/**
 * Freighter Wallet connector for Stellar blockchain.
 * Implements Stellar wallet functionality using the Freighter browser extension.
 */
class FreighterWalletConnector extends StellarWalletConnector {
    constructor(opts) {
        super('Freighter', opts);
        this.name = 'Freighter';
        this.overrideKey = 'freighter';
    }
    getProvider() {
        if (!this.isInstalledOnBrowser()) {
            return undefined;
        }
        if (!this.provider) {
            this.provider = createFreighterProvider();
        }
        return this.provider;
    }
    isInstalledOnBrowser() {
        return FreighterWalletConnector.isInstalled;
    }
    static isConnected() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const response = yield isConnected();
            const connected = (_a = response === null || response === void 0 ? void 0 : response.isConnected) !== null && _a !== void 0 ? _a : false;
            FreighterWalletConnector.isInstalled = connected;
            return connected;
        });
    }
}
FreighterWalletConnector.isInstalled = false;

export { FreighterWalletConnector };
